File: api/_lib/apex-blanks.ts
/**
 * Apex Fill-in-the-Blanks generation and evaluation (LLM-based).
 */
import crypto from 'node:crypto';
import { generateJSON } from './openai';

export interface ApexBlankItem {
  blank_id: string;
  placeholder: string;
  expected_answer: string;
  guidance?: string;
}

export interface ApexFillInBlankQuestion {
  id: string;
  title: string;
  instructions: string;
  code_with_blanks: string;
  blanks: ApexBlankItem[];
  points: number;
  topic: string;
  difficulty: 'easy' | 'medium' | 'hard';
}

export interface ApexBlanksEvaluationResult {
  question_id: string;
  score: number; // 0..points
  max_score: number;
  feedback: string;
  per_blank: Array<{
    blank_id: string;
    correct: boolean;
    expected: string;
    received: string;
    notes?: string;
  }>;
}

function stableId(prefix: string): string {
  return `${prefix}_${crypto.randomUUID()}`;
}

export async function generateApexFillInTheBlanks(opts: {
  job: {
    title?: string | null;
    role?: string | null;
    level?: string | null;
    description?: string | null;
    must_have_skills?: string[] | null;
    good_to_have_skills?: string[] | null;
  };
  count: number;
  difficulty: string;
}): Promise<ApexFillInBlankQuestion[]> {
  const count = Math.max(1, Math.min(20, Number(opts.count) || 5));
  const difficulty = String(opts.difficulty || 'medium').toLowerCase();

  const jobText = [
    opts.job.title,
    opts.job.role,
    opts.job.level,
    opts.job.description,
    ...(opts.job.must_have_skills || []),
    ...(opts.job.good_to_have_skills || []),
  ]
    .filter(Boolean)
    .join(' ')
    .slice(0, 1200);

  const prompt = `
You are an expert Salesforce technical assessor.
Generate exactly ${count} Apex "Fill in the Blanks" questions.

Context (Salesforce role): ${jobText}
Difficulty: ${difficulty}

STRICT REQUIREMENTS:
- These are NOT DSA problems.
- Each question must be based on practical Salesforce/Apex knowledge.
- Each question must contain EXACTLY ONE blank.
- In the code snippet, mark the blank location with the EXACT marker: /* BLANK */
- The marker /* BLANK */ must appear EXACTLY ONCE in code_with_blanks.
- Do NOT use placeholders like [[blank]] or [[BLANK_1]] in the question text or code.
- Do NOT provide hints, guidance, or suggestions for the blank.
- The blank should test syntax and platform concepts: SOQL, DML, triggers, bulkification, governor limits, sharing, exception handling, collections, null checks.
- Keep snippets small enough to read quickly (roughly 15-40 lines).
- For the blank, provide the EXACT expected answer string (case-sensitive).

OUTPUT JSON ONLY:
{
  "questions": [
    {
      "title": "...",
      "instructions": "...",
      "code_with_blanks": "...",
      "topic": "SOQL|Triggers|Governor Limits|Security|DML|Collections|Async|Testing|General",
      "difficulty": "easy|medium|hard",
      "points": 10,
      "blanks": [
        {
          "blank_id": "BLANK_1",
          "placeholder": "",
          "expected_answer": "...",
          "guidance": ""
        }
      ]
    }
  ]
}
`.trim();

  const result = await generateJSON<any>(prompt, {
    temperature: 0.7,
    maxTokens: Math.min(7000, 1600 + count * 700),
    timeoutMs: 16000,
  });

  const raw: any[] = Array.isArray(result?.questions) ? result.questions : [];
  const normalized: ApexFillInBlankQuestion[] = raw
    .map((q: any, i: number) => {
      const blanks: ApexBlankItem[] = Array.isArray(q?.blanks)
        ? q.blanks
            .map((b: any) => ({
              blank_id: String(b?.blank_id || '').trim() || `BLANK_${i + 1}`,
              placeholder: String(b?.placeholder || '').trim() || '',
              expected_answer: String(b?.expected_answer || '').trim(),
              guidance: String(b?.guidance || '').trim() || undefined,
            }))
            .filter((b: ApexBlankItem) => b.expected_answer.length > 0)
        : [];

      const code = String(q?.code_with_blanks || '').trim();
      if (!code || blanks.length < 1) return null;

      const markerCount = (code.match(/\/\*\s*BLANK\s*\*\//g) || []).length;
      if (markerCount !== 1) return null;

      const firstBlank = blanks[0];
      const normalizedBlanks: ApexBlankItem[] = [
        {
          blank_id: String(firstBlank.blank_id || 'BLANK_1').trim() || 'BLANK_1',
          placeholder: '',
          expected_answer: String(firstBlank.expected_answer || '').trim(),
        },
      ].filter((b) => b.expected_answer.length > 0);

      if (normalizedBlanks.length !== 1) return null;

      return {
        id: stableId('apex_blank'),
        title: String(q?.title || `Apex Fill-in-the-Blanks ${i + 1}`).trim(),
        instructions: String(q?.instructions || 'Fill in the missing Apex syntax.').trim(),
        code_with_blanks: code,
        blanks: normalizedBlanks,
        points: typeof q?.points === 'number' ? q.points : 10,
        topic: String(q?.topic || 'General').trim(),
        difficulty: (String(q?.difficulty || difficulty).toLowerCase() as any) || 'medium',
      };
    })
    .filter((q): q is ApexFillInBlankQuestion => Boolean(q));

  return normalized.slice(0, count);
}

export async function evaluateApexFillInTheBlanks(opts: {
  questions: ApexFillInBlankQuestion[];
  submissions: Record<string, Record<string, string>>; // question_id -> blank_id -> answer
}): Promise<{ results: ApexBlanksEvaluationResult[]; total_score: number; max_score: number }> {
  const questions = opts.questions || [];
  const submissions = opts.submissions || {};

  const maxScore = questions.reduce((sum, q) => sum + (Number(q.points) || 0), 0);

  const prompt = `
You are a strict Apex syntax evaluator.
Evaluate the candidate's answers for Apex fill-in-the-blanks questions.

RULES:
- Grade each blank as correct/incorrect.
- Correctness requires exact syntax (case-sensitive) unless Apex is case-insensitive for that token (e.g., keywords). Use reasonable judgment but be strict.
- If minor whitespace differences only, consider correct.
- Provide constructive feedback.
- Score each question between 0 and its points.

INPUT JSON:
${JSON.stringify({ questions, submissions })}

OUTPUT JSON ONLY:
{
  "results": [
    {
      "question_id": "...",
      "score": 0,
      "max_score": 10,
      "feedback": "...",
      "per_blank": [
        {"blank_id": "BLANK_1", "correct": true, "expected": "...", "received": "...", "notes": "..."}
      ]
    }
  ]
}
`.trim();

  const judged = await generateJSON<any>(prompt, {
    temperature: 0.2,
    maxTokens: 12000,
    timeoutMs: 25000,
  });

  const results: ApexBlanksEvaluationResult[] = Array.isArray(judged?.results) ? judged.results : [];

  const total = results.reduce((sum, r) => sum + (Number(r.score) || 0), 0);
  return { results, total_score: total, max_score: maxScore };
}

File: backend/app/services/ai/apex_blanks.py
from __future__ import annotations

from typing import Any, Dict, List

from app.services.ai.factory import get_ai


async def generate_apex_fill_in_the_blanks(
    *,
    job: Dict[str, Any],
    count: int,
    difficulty: str,
) -> List[Dict[str, Any]]:
    ai = get_ai()

    prompt = (
        "You are an expert Salesforce Apex interviewer."
        " Create Apex fill-in-the-blanks questions for a candidate assessment.\n\n"
        "JOB CONTEXT JSON:\n"
        f"{job}\n\n"
        "REQUIREMENTS:\n"
        f"- Generate exactly {count} questions.\n"
        f"- Difficulty: {difficulty}.\n"
        "- Each question must include a realistic Apex code snippet with blanks removed.\n"
        "- CRITICAL: Replace each missing part in the code with the EXACT token ___ (three underscores).\n"
        "- The code field MUST contain ___ for every blank, in the same order as the blanks array.\n"
        "- Example code: 'List<Account> accs = [___ Id, Name FROM ___ LIMIT ___];'\n"
        "- Each blank must have a unique id (BLANK_1, BLANK_2, ...) in order of appearance in the code.\n"
        "- placeholder must be a short descriptive hint (e.g. 'SOQL keyword', 'object name', 'integer limit').\n"
        "- Provide expected answers for each blank.\n"
        "- Include points per question (5-15).\n\n"
        "OUTPUT JSON ONLY in this shape:\n"
        "{\n"
        '  "questions": [\n'
        "    {\n"
        '      "id": "q1",\n'
        '      "title": "SOQL Query",\n'
        '      "prompt": "Complete the SOQL query to fetch accounts.",\n'
        '      "code": "List<Account> accs = [___ Id FROM ___ LIMIT ___];",\n'
        '      "blanks": [\n'
        '        {"blank_id": "BLANK_1", "placeholder": "SOQL keyword", "expected": "SELECT"},\n'
        '        {"blank_id": "BLANK_2", "placeholder": "object name", "expected": "Account"},\n'
        '        {"blank_id": "BLANK_3", "placeholder": "integer limit", "expected": "100"}\n'
        "      ],\n"
        '      "points": 10\n'
        "    }\n"
        "  ]\n"
        "}"
    )

    generated = await ai.generate_json(prompt, temperature=0.4, max_tokens=6000, timeout_s=45)
    questions = generated.get("questions") if isinstance(generated, dict) else None
    if not isinstance(questions, list):
        return []
    result = []
    for q in questions:
        if not isinstance(q, dict):
            continue
        # Normalize field names to match frontend interface
        if "code_with_blanks" not in q and "code" in q:
            q["code_with_blanks"] = q["code"]
        if "instructions" not in q and "prompt" in q:
            q["instructions"] = q["prompt"]
        if "instructions" not in q and "title" in q:
            q["instructions"] = q["title"]
        if "code_with_blanks" not in q:
            q["code_with_blanks"] = ""
        # Normalize blank guidance field
        for blank in q.get("blanks", []):
            if isinstance(blank, dict) and "guidance" not in blank:
                blank["guidance"] = blank.get("hint") or blank.get("notes") or ""
        result.append(q)
    return result


async def evaluate_apex_fill_in_the_blanks(
    *,
    questions: List[Dict[str, Any]],
    submissions: Dict[str, Dict[str, str]],
) -> Dict[str, Any]:
    """Port of Node.js evaluateApexFillInTheBlanks() logic.

    Input:
    - questions: list of apex blanks questions stored in session.proctoring_data.assessment_content.apex_blanks
    - submissions: question_id -> blank_id -> answer

    Output:
    - { results: [...], total_score: number, max_score: number }
    """

    max_score = sum(float(q.get("points") or 0) for q in (questions or []))

    prompt = (
        "You are a strict Apex syntax evaluator.\n"
        "Evaluate the candidate's answers for Apex fill-in-the-blanks questions.\n\n"
        "RULES:\n"
        "- Grade each blank as correct/incorrect.\n"
        "- Correctness requires exact syntax (case-sensitive) unless Apex is case-insensitive for that token (e.g., keywords). Use reasonable judgment but be strict.\n"
        "- If minor whitespace differences only, consider correct.\n"
        "- Provide constructive feedback.\n"
        "- Score each question between 0 and its points.\n\n"
        "INPUT JSON:\n"
        f"{ {'questions': questions, 'submissions': submissions} }\n\n"
        "OUTPUT JSON ONLY:\n"
        "{\n"
        '  "results": [\n'
        "    {\n"
        '      "question_id": "...",\n'
        '      "score": 0,\n'
        '      "max_score": 10,\n'
        '      "feedback": "...",\n'
        '      "per_blank": [\n'
        '        {"blank_id": "BLANK_1", "correct": true, "expected": "...", "received": "...", "notes": "..."}\n'
        "      ]\n"
        "    }\n"
        "  ]\n"
        "}\n"
    )

    ai = get_ai()
    judged = await ai.generate_json(prompt, temperature=0.2, max_tokens=4000, timeout_s=60)

    results = judged.get("results") if isinstance(judged, dict) else None
    if not isinstance(results, list):
        results = []

    total_score = 0.0
    for r in results:
        if isinstance(r, dict):
            try:
                total_score += float(r.get("score") or 0)
            except Exception:
                pass

    return {
        "results": results,
        "total_score": total_score,
        "max_score": max_score,
    }
